src/Vulkan/VMASharp_API_For_Review.md

```diff
@@ -0,0 +1,307 @@
+  {
+    public const long LostAllocationCost = 1048576;
+    public long Offset;
+    public long SumFreeSize;
+    public long SumItemSize;
+    public long ItemsToMakeLostCount;
+    public object Item;
+    public object CustomData;
+    public AllocationRequestType Type;
+    public readonly long CalcCost();
+  }
+  public enum AllocationRequestType
+  {
+    Normal,
+    UpperAddress,
+    EndOfList1,
+    EndOfList2,
+  }
+  [Flags]
+  public enum AllocationStrategyFlags
+  {
+    BestFit = 1,
+    WorstFit = 2,
+    FirstFit = 4,
+    MinMemory = BestFit, // 0x00000001
+    MinTime = FirstFit, // 0x00000004
+    MinFragmentation = WorstFit, // 0x00000002
+  }
+  [Flags]
+  public enum AllocatorCreateFlags
+  {
+    ExternallySyncronized = 1,
+    ExtMemoryBudget = 8,
+    AMDDeviceCoherentMemory = 16, // 0x00000010
+    BufferDeviceAddress = 32, // 0x00000020
+  }
+  public sealed class BlockAllocation : Allocation
+  {
+    public override DeviceMemory DeviceMemory { get; }
+    public override long Offset { get; internal set; }
+    public override IntPtr MappedData { get; }
+    public override IntPtr Map();
+    public override void Unmap();
+  }
+  public class MapMemoryException : VulkanResultException
+  {
+    public MapMemoryException(string message);
+    public MapMemoryException(Result res);
+    public MapMemoryException(string message, Result res);
+  }
+  public enum MemoryUsage
+  {
+    Unknown,
+    GPU_Only,
+    CPU_Only,
+    CPU_To_GPU,
+    GPU_To_CPU,
+    CPU_Copy,
+    GPU_LazilyAllocated,
+  }
+  [Flags]
+  public enum PoolCreateFlags
+  {
+    IgnoreBufferImageGranularity = 1,
+    LinearAlgorithm = 16, // 0x00000010
+    BuddyAlgorithm = 32, // 0x00000020
+  }
+  public struct PoolStats
+  {
+    public long Size;
+    public long UnusedSize;
+    public int AllocationCount;
+    public int UnusedRangeCount;
+    public long UnusedRangeSizeMax;
+    public int BlockCount;
+  }
+  public struct StatInfo
+  {
+    public int BlockCount;
+    public int AllocationCount;
+    public int UnusedRangeCount;
+    public long UsedBytes;
+    public long UnusedBytes;
+    public long AllocationSizeMin;
+    public long AllocationSizeAvg;
+    public long AllocationSizeMax;
+    public long UnusedRangeSizeMin;
+    public long UnusedRangeSizeAvg;
+    public long UnusedRangeSizeMax;
+  }
+  public class Stats
+  {
+    public readonly StatInfo[] MemoryType;
+    public readonly StatInfo[] MemoryHeap;
+    public StatInfo Total;
+  }
+  public enum SuballocationType
+  {
+    Free,
+    Unknown,
+    Buffer,
+    Image_Unknown,
+    Image_Linear,
+    Image_Optimal,
+  }
+  public class ValidationFailedException : ApplicationException
+  {
+  }
+  public sealed class VulkanMemoryAllocator : IDisposable
+  {
+    public Device Device { get; }
+    public VulkanMemoryAllocator(in VulkanMemoryAllocatorCreateInfo createInfo);
+    public int CurrentFrameIndex { get; set; }
+    public void Dispose();
```

Comment:
Does the allocator dispose all memory pools too ?

Reply:
I believe the current implementation throws if there's memory still allocated when you try to dispose it. But we can do whatever here as we are moving away from VMA altogether.